File: examples/kotlin/sanitizer_test/Handlers.kt
// Kotlin sanitizer-fixture — mirrors the Java fixture.
package sanitizer_test

import java.sql.Connection
import org.springframework.web.util.HtmlUtils

class Handlers(private val conn: Connection) {

    fun sqlRaw(userId: String) =
        conn.createStatement().executeQuery("SELECT * FROM users WHERE id = '$userId'")

    fun sqlSafe(userId: String): java.sql.ResultSet {
        val stmt = conn.prepareStatement("SELECT * FROM users WHERE id = ?")
        stmt.setString(1, userId)
        return stmt.executeQuery()
    }

    fun xssRaw(name: String): String = "<p>Hello, $name</p>"

    fun xssSafe(name: String): String {
        val safe = HtmlUtils.htmlEscape(name)
        return "<p>Hello, $safe</p>"
    }
}

File: test-fixtures/languages/php/sanitizer_test/handlers.php
<?php
// PHP sanitizer-fixture — parallel handlers per sink family.

class Handlers {
    private $pdo;

    public function __construct($pdo) {
        $this->pdo = $pdo;
    }

    // --- Command injection ----------------------------------------------

    public function cmdRaw($input) {
        return exec("ping " . $input);
    }

    public function cmdSafe($input) {
        $safe = escapeshellarg($input);
        return exec("ping " . $safe);
    }

    // --- SQL injection --------------------------------------------------

    public function sqlRaw($userId) {
        return $this->pdo->query("SELECT * FROM users WHERE id = '" . $userId . "'");
    }

    public function sqlSafe($userId) {
        $stmt = $this->pdo->prepare("SELECT * FROM users WHERE id = ?");
        $stmt->execute([$userId]);
        return $stmt;
    }

    // --- XSS ------------------------------------------------------------

    public function xssRaw($name) {
        return "<p>Hello, " . $name . "</p>";
    }

    public function xssSafe($name) {
        $safe = htmlspecialchars($name, ENT_QUOTES, 'UTF-8');
        return "<p>Hello, " . $safe . "</p>";
    }
}
